### PR TITLE
[SHLO] [TTNN] Add users-dim factor to paged ops + force MeshWorkloadFactory for DP

### PR DESCRIPTION
**Pull request by alinakhanTT**
## Summary
- Add users-dim factor to Shardy custom sharding rules for `paged_update_cache`, `paged_fill_cache`, and `paged_scaled_dot_product_attention_decode` so the SPMD partitioner can keep each DP replica's cache update local (KV cache corruption across replicas otherwise).
- Force `MeshWorkloadFactory` for `paged_update_cache` and `paged_fill_cache` by passing the cache tensor's full set of mesh coords, so the mesh-aware factory is always selected under DP partitioning.
- Drop unused `buildHeadShardedCustomCallRule` helper (all three paged ops now inline their builders).

Co-authored with @ssaliceTT, who rebased on top of the tt-mlir commit pinned by tt-xla `main` (`464a5f341`) and added the `paged_fill_cache` rule update.

## Test plan
- Exercised by tt-xla DP+TP integration tests on n300-llmbox (tt-xla PR tenstorrent/tt-xla#4947, CI green against this tt-mlir tree).
- tt-mlir unit tests pass.

### CODE REVIEW

File: lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp [hunk 1]
```diff
@@ -343,24 +323,48 @@ getPagedAttentionShardingRule(mlir::stablehlo::CustomCallOp op) {
...
+    // table / cur_pos sharding can't propagate and a single replica observes
+    // KV cache corruption from the other replica's writes.
+    SmallVector<int64_t> usersOperandDims(numOperands, mlir::sdy::kNullDim);
+    SmallVector<int64_t> usersResultDims(numResults, mlir::sdy::kNullDim);
+    usersOperandDims[0] = queryUsersDim; // query
+    if (numOperands > 3) {
+      usersOperandDims[3] = 0; // page_table
+    }
+    if (numOperands > 4) {
+      usersOperandDims[4] = 0; // cur_pos_tensor
```
Comment (sshonTT) [inline lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp]: I think we need a guard to index cur_pos_tensor. usersOperandDims[4] could be a attention_mask. Please check https://github.com/tenstorrent/tt-xla/blob/main/python_package/tt_torch/custom_ops.py#L826-L835

Comment (alinakhanTT) [reply lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp]: Fixed in 4652d8dd6!

**Review by sshonTT — APPROVED** [review]
looks good for sharding rule part.

File: runtime/lib/ttnn/operations/kv_cache/paged_fill_cache.cpp [hunk 2]
```diff
@@ -27,10 +27,18 @@ void run(const ::tt::target::ttnn::PagedFillCacheOp *op,
...
 
+  // Force the MeshWorkloadFactory by passing the cache tensor's full set of
+  // mesh coordinates. With std::nullopt, the device op selects the
+  // single-program factory, which under DP partitioning runs the kernel on
+  // the mesh root only - so per-rank cache writes from non-root devices are
+  // silently dropped.
+  const auto &coordVec = cacheTensor.tensor_topology().mesh_coords();
+  std::set<::ttnn::MeshCoordinate> meshCoords(coordVec.begin(), coordVec.end());
+
   ::ttnn::experimental::paged_fill_cache(cacheTensor, inputTensor,
```
Comment (pilkicTT) [inline runtime/lib/ttnn/operations/kv_cache/paged_fill_cache.cpp:42]: shouldn't `ttnn` do this by default? is there a reason why we are explicitly passing this - since we are passing the `cacheTensor` as input of the op, they could do this on their end...

maybe we should file an issue to them (if we haven't already)...

**Review by pilkicTT — APPROVED** [review]

File: lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp [hunk 3]
```diff
@@ -602,24 +602,73 @@ getPagedAttentionShardingRule(mlir::stablehlo::CustomCallOp op) {
...
+    SmallVector<int64_t> usersResultDims(numResults, mlir::sdy::kNullDim);
+    usersOperandDims[0] = queryUsersDim; // query
+
+    // Operand layout (see tt_torch/custom_ops.py): the base operands are
+    // [query, key, value, page_table]; an optional attention_mask, then an
+    // optional cur_pos_tensor, then an optional attention_sink follow. Only
+    // page_table (fixed at index 3) and cur_pos_tensor carry the users/batch
+    // dim, so we must not assume cur_pos_tensor is at index 4 -- with an
+    // attention_mask present, index 4 is the mask and cur_pos_tensor shifts
+    // to 5. Use the has_* frontend flags to locate it.
```
Comment (mmanzoorTT) [inline lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp]: Refer op definition in [TTIROps.td](https://github.com/tenstorrent/tt-mlir/blob/main/include/ttmlir/Dialect/TTIR/IR/TTIROps.td#L6112) instead of `tt_torch/custom_ops.py`;

**Review by mmanzoorTT — COMMENTED** [review]
Please add a test for these changes.

File: runtime/lib/ttnn/operations/kv_cache/paged_fill_cache.cpp (hunk 2 above)
Comment (alinakhanTT) [reply runtime/lib/ttnn/operations/kv_cache/paged_fill_cache.cpp:42]: Agreed it would be cleaner for ttnn to infer this from the cache tensor. Today I pass `mesh_coords` explicitly because with `std::nullopt` the device op selects the single-program factory, which under DP partitioning runs the kernel on the mesh root only, so per-rank cache writes from non-root devices get silently dropped (KV cache corruption). Forcing the `MeshWorkloadFactory` via the cache tensor’s full coord set is the workaround

File: lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp (hunk 3 above)
Comment (alinakhanTT) [reply lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp]: Done!

File: runtime/lib/ttnn/operations/kv_cache/paged_fill_cache.cpp (hunk 2 above)
Comment (alinakhanTT) [reply runtime/lib/ttnn/operations/kv_cache/paged_fill_cache.cpp:42]: Issue filed: [tenstorrent/tt-metal#47955](https://github.com/tenstorrent/tt-metal/issues/47955)

**Review by mmanzoorTT — APPROVED** [review]
